pydantic/README.rst

```diff
@@ -0,0 +1,201 @@
+    :target: https://odoo-community.org/page/development-status
+    :alt: Beta
+.. |badge2| image:: https://img.shields.io/badge/licence-LGPL--3-blue.png
+    :target: http://www.gnu.org/licenses/lgpl-3.0-standalone.html
+    :alt: License: LGPL-3
+.. |badge3| image:: https://img.shields.io/badge/github-OCA%2Frest--framework-lightgray.png?logo=github
+    :target: https://github.com/OCA/rest-framework/tree/14.0/pydantic
+    :alt: OCA/rest-framework
+.. |badge4| image:: https://img.shields.io/badge/weblate-Translate%20me-F47D42.png
+    :target: https://translation.odoo-community.org/projects/rest-framework-14-0/rest-framework-14-0-pydantic
+    :alt: Translate me on Weblate
+.. |badge5| image:: https://img.shields.io/badge/runbot-Try%20me-875A7B.png
+    :target: https://runbot.odoo-community.org/runbot/271/14.0
+    :alt: Try me on Runbot
+
+|badge1| |badge2| |badge3| |badge4| |badge5| 
+
+This addon allows you to define inheritable `Pydantic classes <https://pydantic-docs.helpmanual.io/>`_.
+
+**Table of contents**
+
+.. contents::
+   :local:
+
+Usage
+=====
+
+To define your own pydantic model you just need to create a class that inherits from
+``odoo.addons.pydantic.models.BaseModel`` or a subclass of.
+
+.. code-block:: python
+
+    from odoo.addons.pydantic.models import BaseModel
+    from pydantic import Field
+
+
+    class PartnerShortInfo(BaseModel):
+        id: str
+        name: str
+
+
+    class PartnerInfo(BaseModel):
+        street: str
+        street2: str = None
+        zip_code: str = None
+        city: str
+        phone: str = None
+        is_componay : bool = Field(None)
+
+
+In the preceding code, 2 new models are created, one for each class. If you
+want to extend an existing model, you must pass the extended pydantic model
+trough the `extends` parameter on class declaration.
+
+.. code-block:: python
+
+    class Coordinate(models.BaseModel):
+        lat = 0.1
+        lng = 10.1
+
+    class PartnerInfoWithCoordintate(PartnerInfo, extends=PartnerInfo):
+        coordinate: Coordinate = None
+
+`PartnerInfoWithCoordintate` extends `PartnerInfo`. IOW, Base class are now the
+same and define the same fields and methods. They can be used indifferently into
+the code. All the logic will be provided by the aggregated class.
+
+.. code-block:: python
+
+    partner1 = PartnerInfo.construct()
+    partner2 = PartnerInfoWithCoordintate.construct()
+
+    assert partner1.__class__ == partner2.__class__
+    assert PartnerInfo.schema() == PartnerInfoWithCoordintate.schema()
+
+.. note::
+
+    Since validation occurs on instance creation, it's important to avoid to
+    create an instance of a Pydantic class by usign the normal instance
+    constructor `partner = PartnerInfo(..)`. In such a case, if the class is
+    extended by an other addon and a required field is added, this code will
+    no more work. It's therefore a good practice to use the `cosntruct()` class
+    method to create a pydantic instance.
```

Comment:
I'd add an mention that adding required fields violates de Liskov substitution principle and should generally be avoided. This is certainly forbidden in requests data structures. When extending response data structures I can see why this is useful, to document new fields that are guaranteed to be present when extension addons are installed.

I wonder if we can collect example use cases for extending data structures.

Reply:
> I'd add an mention that adding required fields violates de Liskov substitution principle and should generally be avoided. This is certainly forbidden in requests data structures. When extending response data structures I can see why this is useful, to document new fields that are guaranteed to be present when extension addons are installed.

:+1: 

> I wonder if we can collect example use cases for extending data structures.

We could but it's a tedious work